pep-9999.rst

```diff
@@ -85,7 +85,112 @@ Supported Constructs
+construct as they best see fit or report an error. Linters should usually
+flag those constructs. Type stub authors should avoid using them.
+
+Imports
+-------
+
+Type stubs distinguish between imports that are re-exported and those
+that are only used internally. Imports are only re-exported if they
+use one of these forms:[#pep484]_
+
+* `import ... as ...`
+* `from ... import as ...`
+* `from ... import *`
+
+Stub authors can use `import foo as foo` to re-export an import using
+the same name::
+
+    from foo import func as func  # re-exported as "func"
+    from foo import Cls as OtherCls  # re-exported as "OtherCls"
+    from foo import baz  # internal use only
+    import bar as bar  # re-exported as "bar"
+    import baz  # internal use only
+
+Sub-modules are always exported when they are imported in a module.
+For example, consider the following file structure::
+
+    foo/
+        __init__.pyi
+        bar.pyi
+
+Then `foo` will export `bar` when one of the following constructs is used in
+`__init__.pyi`::
+
+    from . import bar
+    from .bar import Bar
+
+Module Level Attributes
+-----------------------
+
+Module level variables and constants can be annotated using either
+type comments or variable annotation syntax::
+
+    x: int  # recommended
+    x: int = 0
+    x = 0  # type: int
+    x = ...  # type: int
+
+The type of an unannotated variable is unspecified. Same when the annotation
+and assigned value disagree. As an exception, the ellipsis literal can
+stand in for any type::
+
+    x = 0  # type is unspecified
+    x: int = ""  # type is unspecified
+    x: int = ...  # type is int
+
+Functions and Methods
+---------------------
+
+Function and method definition syntax follows general Python syntax.
+Unless an argument is prefixed with two underscores (but not suffixed
+with two underscores), it can be used as a keyword argument[#pep484]_::
+
+    # x is positional-only
+    # y can be used positionally or as keyword argument
+    # z is keyword-only
+    def foo(__x, y, *, z): ...
+
+If an argument is unannotated, its type is assumed to be `Any`. The type of
+an argument where annotation and the type of the default argument disagree
+is unspecified. As an exception, the ellipsis literal can stand in for any
+type::
+
+    # x has type Any
+    # y's type is unspecified
+    # z has type int
+    def foo(x, y: int = "", z: int = ...): ...
+
+For a class `C`, the type of the first argument to a classmethod is
+assumed to be `Type[C]`, if unannotated. For other non-static methods,
+it type is assumed to be `C`::
+
+    class Foo:
+        def do_things(self): ...  # self has type Foo
+        @classmethod
+        def create_it(cls): ...  # cls has type Type[Foo]
+        @staticmethod
+        def utility(x): ...  # x has type Any
+
+But::
+
+    _T = TypeVar("_T")
+
+    class Foo:
+        def do_things(self: _T): ...  # self has type _T
+        @classmethod
+        def create_it(cls: _T): ...  # cls has type _T
```

Comment:
Should we say something about how this is only useful if `_T` shows up in other annotations on this method, and recommend against explicit annotations otherwise?

Reply:
In the style section we already have the following recommendation:

> Do not annotate ``self`` and ``cls`` in method definitions, except when referencing a type variable.

Is this sufficient?